watcher/watcher.go

```diff
@@ -600,75 +588,208 @@ func (w *Watcher) checkState(
 	roomCount *models.RoomsStatusCount,
 	scheduler *models.Scheduler,
 	nowTimestamp int64,
-) (bool, bool, bool, error) { //shouldScaleUp, shouldScaleDown, changedState
+) (int, bool, bool, bool, error) { //delta, shouldScaleUp, shouldScaleDown, changedState
 	var shouldScaleUp, shouldScaleDown, changedState bool
+	var delta int
+	var err error
 	inSync := true
 
+	// Creating or Terminating state
 	if scheduler.State == models.StateCreating || scheduler.State == models.StateTerminating {
-		return false, false, changedState, nil
+		return 0, false, false, changedState, nil
 	}
+	// Rooms below Min
 	if roomCount.Total() < autoScalingInfo.Min {
-		return true, false, changedState, nil
+		return 0, true, false, changedState, nil
 	}
-	if autoScalingInfo.Min > 0 && 100*roomCount.Occupied >= roomCount.Total()*autoScalingInfo.Up.Trigger.Limit {
-		return true, false, changedState, nil
+	// Rooms above Max
+	if autoScalingInfo.Max > 0 && roomCount.Total() > autoScalingInfo.Max {
+		return 0, false, true, changedState, nil
 	}
 
-	threshold := autoScalingInfo.Up.Trigger.Threshold
-	usage := float32(autoScalingInfo.Up.Trigger.Usage) / 100
-	capacity := autoScalingInfo.Up.Trigger.Time / w.AutoScalingPeriod
+	// Up
+	if (roomCount.Total() < autoScalingInfo.Max && autoScalingInfo.Max > 0) || autoScalingInfo.Max == 0 {
+		if len(autoScalingInfo.Up.MetricsTrigger) > 0 {
+			delta, shouldScaleUp, changedState, inSync, err = w.checkMetricsTrigger(autoScalingInfo, autoScalingInfo.Up, w.ScaleUpInfo, roomCount, scheduler, nowTimestamp)
+		} else {
+			delta, shouldScaleUp, changedState, inSync, err = w.checkLegacyTrigger(autoScalingInfo, autoScalingInfo.Up, w.ScaleUpInfo, roomCount, scheduler, nowTimestamp)
+		}
+	}
 
-	isAboveThreshold, err := w.ScaleUpInfo.SendUsageAndReturnStatus(
-		w.SchedulerName,
-		capacity, roomCount.Occupied, roomCount.Total(),
-		threshold, usage,
-	)
-	if err != nil {
-		return false, false, false, err
+	if shouldScaleUp == false && inSync == true && roomCount.Total() > autoScalingInfo.Min {
+		// Down
+		if len(autoScalingInfo.Down.MetricsTrigger) > 0 {
+			delta, shouldScaleDown, changedState, inSync, err = w.checkMetricsTrigger(autoScalingInfo, autoScalingInfo.Down, w.ScaleDownInfo, roomCount, scheduler, nowTimestamp)
+		} else {
+			delta, shouldScaleDown, changedState, inSync, err = w.checkLegacyTrigger(autoScalingInfo, autoScalingInfo.Down, w.ScaleDownInfo, roomCount, scheduler, nowTimestamp)
+		}
+	} else if inSync == true { // always send current usage
+		if len(autoScalingInfo.Down.MetricsTrigger) > 0 {
+			for _, trigger := range autoScalingInfo.Down.MetricsTrigger {
+				w.ScaleDownInfo.SendUsage(
```

Comment:
I couldn't find any sendUsage for scale up, it this intended? or am I missing something?

Reply:
On scaleUp it will call SendUsageAndReturnStatus(). SendUsage() is called here because checkMetricsTrigger will not be called for down scaling if up scaling is needed.

Reply:
so we have two separate usage metrics, one for scaling up and another for scaling down?